File: 算法/位运算/剑指offer15二进制中1的个数.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution
{
public:
    int hammingWeight(uint32_t m)
    {
        // 因为这是无符号整数
        // 直接右移即可
        int cnt = 0;
        while (m)
        {
            if (m & 1)
            {
                cnt++;
            }
            m = m >> 1;
        }
        return cnt;
    }
};

File: 算法/位运算/剑指offer15二进制中1的个数.py
# python中没有无符号整数的概念


class Solution:
    def hammingWeight(self, n: int) -> int:
        flag = 1  # 用flag左移
        cnt = 0
        max_flag = 1 << 32
        # python中的整数可以超出32位的

        while flag < max_flag:
            if n & flag:
                cnt += 1
            flag = flag << 1

        return cnt


def main():
    a = 0b00000000000000000000000000001011
    s = Solution()
    res = s.hammingWeight(a)
    print(res)


if __name__ == "__main__":
    main()
